### PR TITLE
initial work on danish.

### PR DESCRIPTION
**Pull request by andlo**
## How to test
difficult to test without translating some basic skills - fallback-unknown is one important one…..

## Contributor license agreement signed?
CLA [ X ] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

### CODE REVIEW

File: mycroft/util/parse.py [hunk 1]
```diff
@@ -22,6 +23,11 @@
 from mycroft.util.lang.parse_es import *
 from mycroft.util.lang.parse_it import *
 from mycroft.util.lang.parse_sv import *
+from mycroft.util.lang.parse_da import *
```
[the remark this answers is not in the document]
Comment (forslund) [reply mycroft/util/parse.py]: You can ignore this comment from codacy

File: mycroft/util/format.py [hunk 2]
```diff
@@ -17,8 +17,12 @@
 from mycroft.util.lang.format_pt import *
 from mycroft.util.lang.format_it import *
 from mycroft.util.lang.format_sv import *
+from mycroft.util.lang.format_da import *
```
[the remark this answers is not in the document]
Comment (forslund) [reply mycroft/util/format.py]: And this one.

**Comment by andlo** [comment]
@JarbasAI I am not sure what you mean by "please add unittest for parse_da and format_da,"
Should I add those to files (by looking at _sv and _de and from that guess what da _da should include ? 

Am I guessing right that I then can run the file and It then test what ever I put inside it ? 

Are there some docs for making this that I have missed reading ?

**Comment by JarbasAl** [comment]
@andlo exactly, those unittests are run automatically by travis on all new Pull Requests, any errors will be detected, this avoids different PR breaking previous code

those unittests will check that your functions produce the expected output for a specific input, you should add a case for every input you expect to parse correctly

you can run those files offline too, very useful for debugging

take a look at any of the other languages unittests if you are unsure how to write them, it's pretty easy, just a bunch of assert statements :)

**Comment by forslund** [comment]
You can run a test-file using *pytest*, `pytest test_parse_se.py` (provided you're in the test/unittests/util folder)

**Comment by forslund** [comment]
Thanks for the update @andlo, Some of the tests are failing, see [Travis](https://travis-ci.org/MycroftAI/mycroft-core/jobs/484851362)

There's also a number of code style issues as reported by pep8speaks above

**Comment by forslund** [comment]
Hi @andlo, it would be nice to be able to include this either in this weeks release or the big 19.02 in two weeks. Do you think you can look over the codacy warnings and pep8 issues? I'll try to do a review tomorrow morning and will give you further feedback then.

**Comment by andlo** [comment]
I can fix the format.py and the test for format I think. But I do have problems with the parse.py. It works, but isnt correct. Most of what I am doing is just copy of the german, and most of it I dont understand fully.

**Comment by andlo** [comment]
I dont know.....I is not easy....I can get the peb8 fixed, and do have most of the test_format working, but parse and test parse is really difficult for me to do. Most of what I do is guessing and trying and looking to the german parsing. And not beeing that good t phyton and programming it is hard work :) but I like it anyway and do try.

**Comment by forslund** [comment]
@andlo, that's what we all do :) And it's really good to get a base to build from.

To make it clearer of what is done and not I think we should just pull in the parts that are mostly working. Would you say that the formatting is pretty Ok but not the parsing?

**Comment by andlo** [comment]
yah formatting are working OK, and test_formatting to. Parse is working (as in dosnt gives errors) but is mostly just copy from german. And test_parse isnt working yet. And parse isnt OK as far as I can tell.

But I can cleanup and do as much as I can. I think I got some time during the week and more time next weekend.

**Comment by forslund** [comment]
Sounds good! Let me know if I can assist in some way.

**Comment by andlo** [comment]
Cleaned up most peb8. 
Still got an error in test_format_da.py and many in test_parse_da.py but that is expected as parse_da.py isnt quite good enough.

**Comment by andlo** [comment]
no more peb8's but still errors inn test_format_da.py and test_parse_da.py.
I will look more into that, but would really like someone else looking to :)

**Comment by andlo** [comment]
Ill gotten most of format_test done, and lot of test_parse done. Making test_parse made me fix lots of issues in parse, so using this test is the way to go :)
But I did have to skip some tests, as I couldnt get it to work. They are remarked out, so posible to get back to it later. Maybe someone other than me could have a look at it.....

**Comment by andlo** [comment]
Cant figure out what travis is complaining about.....

**Comment by krisgesling** [comment]
Hi @Andlo, are you able to see the [output from the build](https://travis-ci.org/MycroftAI/mycroft-core/jobs/493927361)?
If you click on details to the right of the failed test then select one of the build jobs in the table. There are four tests that are run on four different versions of python but they're probably all the same in this case.

There is just one test failing for nice_date_time: 
`AssertionError: 'tirs[26 chars]nuar, to tusind og sytten klokken 2017-01-31 13:22:03' != 'tirs[26 chars]nuar, to tusind og sytten klokken et toogtyve om eftermiddagen'`

It looks like there are a couple of PEP8 errors still as well.

**Comment by andlo** [comment]
> Hi @andlo, are you able to see the output from the build?
Yes I think
> If you click on details to the right of the failed test 
and I am lost. Dont se any details to the right of the failed test.

> There is just one test failing for nice_date_time:
> AssertionError: 'tirs[26 chars]nuar, to tusind og sytten klokken 2017-01-31 13:22:03' != 'tirs[26 chars]nuar, to tusind og sytten klokken et toogtyve om eftermiddagen'

Strange, as running test localy on the device dosnt fail......

**Comment by forslund** [comment]
The failing test can be seen here: https://travis-ci.org/MycroftAI/mycroft-core/jobs/494148032#L1377

It's in the test_format.py (not the test_format_da.py), and seems to be the danish `nice_date_time()`. I think it's that you've added support for expanding the time into words that makes this test fail so it should be updated.

Updating  [/mycroft/res/da-dk/date_time_test.json L30](https://github.com/MycroftAI/mycroft-core/blob/dev/mycroft/res/text/da-dk/date_time_test.json#L30) to

` "1": {"datetime_param": "2017, 1, 31, 13, 22, 3", "now": "None",  "use_24hour": "False", "use_ampm": "True", "assertEqual": "tirsdag, den en og tredivte januar, to tusind og sytten klokken et toogtyve om eftermiddagen"}`

should fix the error.

**Comment by andlo** [comment]
Argh - I m not good at this GitHub.....
Seems I messed up something….Please know I am sorry, and advise what I should do....

**Comment by forslund** [comment]
I'm guessing you merged one or other branch into your work branch. I'll need some time to see how to best untangle this. :)

Welcome to git, this kind of thing is sort of a rite of passage. It happens to us all at some point.

**Comment by andlo** [comment]
thanks - I did merge the 19.02 bransh into my add-danish bransh. Didnt realize what would happen, and stil cant figurewhat i realy should have been dooing....I wanted to have the 19.02 bransh on my device including the Danish translation.....

**Comment by forslund** [comment]
I tried restoring your branch but got some errors after a while. The WIP can be found in the branch

[add-danish-restore](https://github.com/MycroftAI/mycroft-core/tree/add-danish-restore)

I used this guide as base: https://stackoverflow.com/questions/5469619/how-move-all-commits-from-specific-user-to-a-new-branch

**Comment by andlo** [comment]
Wel - Ill guess the easiest is to lose this PR and "start over" by making a new fork and copy files manualy and manual add changes again. That houldnt be to hard.....

And it still has to go to the dev branch I guess

**Comment by andlo** [comment]
Closing this and clean up my mess and try again on a new PR

**Pull request closed without merging**